pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -476,12 +457,24 @@ public void flush() {
     private void flushAsync(ClientCnx cnx) {
         boolean shouldFlush = false;
         if (cumulativeAckFlushRequired) {
-            newMessageAckCommandAndWrite(cnx, consumer.consumerId, lastCumulativeAck.messageId.ledgerId,
-                    lastCumulativeAck.messageId.getEntryId(), lastCumulativeAck.bitSetRecyclable,
+            // `doCumulativeAckAsync` has been invoked before, we need to retrieve the latest value from
+            // `lastCumulativeAck`, here we transfer the ownership of `lastCumulativeAck` to the thread local object
+            // so that we can avoid locking this whole code block or recycling the bit set after `lastCumulativeAck`
+            // is updated.
+            final LastCumulativeAck cumulativeAck = LOCAL_LAST_CUMULATIVE_ACK_TO_FLUSH.get();
+            lastCumulativeAck.moveOwnershipTo(cumulativeAck);
+
+            newMessageAckCommandAndWrite(cnx, consumer.consumerId, cumulativeAck.getMessageId().getLedgerId(),
+                    cumulativeAck.getMessageId().getEntryId(), cumulativeAck.getBitSetRecyclable(),
                     AckType.Cumulative, null, Collections.emptyMap(), false,
                     this.currentCumulativeAckFuture, null);
-            this.consumer.unAckedChunkedMessageIdSequenceMap.remove(lastCumulativeAck.messageId);
+            this.consumer.unAckedChunkedMessageIdSequenceMap.remove(cumulativeAck.getMessageId());
             shouldFlush = true;
+
+            if (!lastCumulativeAck.restoreOwnershipIfEmpty(cumulativeAck)) {
```

Comment:
I was thinking if invoke
https://github.com/apache/pulsar/blob/141c44022a27be2fc07eab9827cfdb168e448953/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java#L561

then restoreOwnershipIfEmpty, it may will lose messages to receive and ack

Reply:
`restoreOwnershipIfEmpty` is called in `flushAsync` and `flushAndClean` will also calls `flushAsync`, do you mean `flushAsync` is invoked concurrently? 

----

But I just found another problem, in this case, the 2nd invocation of `flushAsync` might use a default message id because `lastCumulativeAck` has been reset before by `moveOwnershipTo`.

I think we need to use a CAS operation on `cumulativeAckFlushRequired`. I'll check the code again.

Reply:
I will refactor the existing code again. PTAL after I commit the latest code.